src/lib/AuthManager.js

```diff
@@ -0,0 +1,286 @@
+import browser from './browser-api'
+
+// An AuthSession can be used to make API fetch() calls to Nextcloud. It 
+// overrides the requests headers to reduce the number of times, the server
+// has to validate passwords.
+// In order to do that without interfering with the users session cookies, 
+// it intercepts relevant requests/responses and stores the cookies in a dedicated 
+// CookieJar.
+//
+// Because a CookieJar does not respect any attributes given to 
+// cookies (like "secure" or "expires") and because session timeouts are not
+// detected, the follwing rules apply:
+//  - Keep the lifetime of an AuthSession short!
+//  - Never save an AuthSession to persistent or otherwise insecure memory!
+export class AuthSession {
+
+  // @param "https://example.org" This session can only be used for API calls
+  //          to this server.
+  constructor(serverUrl) {
+    if (serverUrl === undefined) {
+      throw "AuthSession constructor arg undefined: serverUrl"
+    }
+    this.alive = true
+    this.authman = window.authManager
+    this.cookieJar = new CookieJar()
+    this.serverUrl = serverUrl
+    // AuthManager creates an entry in authManager.requests for every request
+    // sent. The entries are deleted if a reponse is received. If no response 
+    // is received, the entries have to be removed by authSession.destroy() 
+    // because AuthManagers are long lived and should not leak memory. 
+    // Therefore an authSession has to know which requests belong to it.
+    this.ownedRequests = {} // dict: request.id => true
+
+    this.needsAuthorizations
+    this.setAuthBuildUpRequired()
+
+    // some browsers might not allow us to acces relevant httpOnly cookies
+    this.everRecievedCookie = false
+  }
+
+  async initialize() {
+    if (this.authman === undefined) {
+      const background = await browser.runtime.getBackgroundPage()
+      this.authman = background.authManager
+      window.authManager = this.authman
+    }
+    this.id = this.authman.newSessionId()
+    this.authman.addSession(this)
+  }
+
+  destructor() {
+    this.authman.removeSession(this)
+    this.alive = false
+  }
+
+  check() {
+    if (!this.alive) {
+      throw "Trying to use an AuthSession wich has been destroyed."
+    }
+  }
+
+  // Mimics behaviour of official fetch, but overwrites certain fields to 
+  // contain the request in this AuthSession. Sets a AuthSessionId in the request.
+  fetch(url, init) {
+    this.check()
+
+    // not quite sure if this an actual fix or just a workaround for this 
+    // specific test: Floccus%20nextcloud-folders%20standard-root%20Sync%20with%20one%20client%20should%20not%20fail%20when%20both%20moving%20folders%20and%20deleting%20their%20contents
```

Comment:
https://docs.nextcloud.com/server/latest/developer_manual//app/requests/controllers.html
 "The session is closed automatically for writing, unless you add the @UseSession annotation!"
PUT may be considered writing by nextcloud and could cause the session to close. 
Hence authBuildUp is required for PUTs.